Mathlib/GroupTheory/Coxeter/Inversion.lean

```diff
@@ -0,0 +1,443 @@
+
+theorem isLeftInversion_mul_right_iff {w : W} :
+    cs.IsLeftInversion (t * w) t ↔ ¬cs.IsLeftInversion w t := by
+  rw [← isRightInversion_inv_iff, ← isRightInversion_inv_iff, mul_inv_rev, ht.inv,
+    ht.isRightInversion_mul_left_iff]
+
+theorem not_isLeftInversion_mul_right_iff {w : W}  :
+    ¬cs.IsLeftInversion (t * w) t ↔ cs.IsLeftInversion w t :=
+  (iff_not_comm.mp ht.isLeftInversion_mul_right_iff).symm
+
+end IsReflection
+
+@[simp]
+theorem isRightInversion_simple_iff_isRightDescent (w : W) (i : B) :
+    cs.IsRightInversion w (s i) ↔ cs.IsRightDescent w i := by
+  simp [IsRightInversion, IsRightDescent, cs.isReflection_simple i]
+
+@[simp]
+theorem isLeftInversion_simple_iff_isLeftDescent (w : W) (i : B) :
+    cs.IsLeftInversion w (s i) ↔ cs.IsLeftDescent w i := by
+  simp [IsLeftInversion, IsLeftDescent, cs.isReflection_simple i]
+
+/-- The right inversion sequence of `ω`. The right inversion sequence of a word
+$s_{i_1} \cdots s_{i_\ell}$ is the sequence
+$$s_{i_\ell}\cdots s_{i_1}\cdots s_{i_\ell}, \ldots,
+    s_{i_{\ell}}s_{i_{\ell - 1}}s_{i_{\ell - 2}}s_{i_{\ell - 1}}s_{i_\ell}, \ldots,
+    s_{i_{\ell}}s_{i_{\ell - 1}}s_{i_\ell}, s_{i_\ell}.$$
+-/
+def rightInvSeq (ω : List B) : List W :=
+  match ω with
+  | []          => []
+  | i :: ω      => (π ω)⁻¹ * (s i) * (π ω) :: rightInvSeq ω
+
+/-- The left inversion sequence of `ω`. The left inversion sequence of a word
+$s_{i_1} \cdots s_{i_\ell}$ is the sequence
+$$s_{i_1}, s_{i_1}s_{i_2}s_{i_1}, s_{i_1}s_{i_2}s_{i_3}s_{i_2}s_{i_1}, \ldots,
+    s_{i_1}\cdots s_{i_\ell}\cdots s_{i_1}.$$
+-/
+def leftInvSeq (ω : List B) : List W :=
+  match ω with
+  | []          => []
+  | i :: ω      => s i :: List.map (MulAut.conj (s i)) (leftInvSeq ω)
+
+local prefix:100 "ris" => cs.rightInvSeq
+local prefix:100 "lis" => cs.leftInvSeq
+
+@[simp] theorem rightInvSeq_nil : ris [] = [] := rfl
+
+@[simp] theorem leftInvSeq_nil : lis [] = [] := rfl
+
+@[simp] theorem rightInvSeq_singleton (i : B) : ris [i] = [s i] := by simp [rightInvSeq]
+
+@[simp] theorem leftInvSeq_singleton (i : B) : lis [i] = [s i] := rfl
+
+theorem rightInvSeq_concat (ω : List B) (i : B) :
+    ris (ω.concat i) = (List.map (MulAut.conj (s i)) (ris ω)).concat (s i) := by
+  induction' ω with j ω ih
+  · simp
+  · dsimp [rightInvSeq]
+    rw [ih]
+    simp only [concat_eq_append, wordProd_append, wordProd_cons, wordProd_nil, mul_one, mul_inv_rev,
+      inv_simple, cons_append, cons.injEq, and_true]
+    group
+
+private theorem leftInvSeq_eq_reverse_rightInvSeq_reverse (ω : List B) :
+    lis ω = (ris ω.reverse).reverse := by
+  induction' ω with i ω ih
+  · simp
+  · rw [leftInvSeq, reverse_cons, ← concat_eq_append, rightInvSeq_concat, ih]
+    simp [map_reverse]
+
+theorem leftInvSeq_concat (ω : List B) (i : B) :
+    lis (ω.concat i) = (lis ω).concat ((π ω) * (s i) * (π ω)⁻¹) := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse, rightInvSeq]
+
+theorem rightInvSeq_reverse (ω : List B) :
```

Comment:
Can (/ should) these be phrased with `Function.LeftInverse`?

Reply:
How?